[FLINK-2569] [core] Add CsvReader support for Value types

chiwanpark opened the pull request:
This PR contains following changes:

* Extend `TupleTypeInfo` to support `Value` type as basic (primitive) type.
* Rename `CsvReaderWithPOJOITCase` to `CsvReaderITCase`
* Add a unit test about type extraction from `CsvReader`
* Add a integrated test about `CsvReader` with `Value` types.

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/typeutils/TupleTypeInfo.java:
```diff
@@ -159,7 +161,8 @@ public String toString() {
-	
+
+	@SuppressWarnings("unchecked")
 	public static <X extends Tuple> TupleTypeInfo<X> getBasicTupleTypeInfo(Class<?>... basicTypes) {
```
I would keep the original method and add a `getBasicAndBasicValueTupleTypeInfo()` method.

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/typeutils/TupleTypeInfo.java:
```diff
@@ -174,13 +177,15 @@ public String toString() {
 			if (info == null) {
-				throw new IllegalArgumentException("Type at position " + i + " is not a basic type.");
+				try {
+					info = ValueTypeInfo.getValueTypeInfo((Class<Value>) type);
```
Any class can implement the `Value` interface.
 We need to check if the class is on of the "BasicValue" classes, that Flink ships.

chiwanpark commented:
@fhueske Thanks for review. :) I addressed your comments.

* Add `getBasicAndBasicValueTupleTypeInfo` method into `TupleTypeInfo`
* Add `isBasicValueType` method into `ValueTypeInfo` class to check whether the type is basic value or not

aljoscha commented:
Looks good to merge :+1:

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/io/CsvReaderITCase.java:
```diff
@@ -123,6 +132,21 @@ public void testPOJOTypeWithFieldsOrderAndFieldsSelection() throws Exception {
+
+		DataSet<Tuple8<StringValue, BooleanValue, ByteValue, ShortValue, IntValue, LongValue, FloatValue, DoubleValue>> data =
+				env.readCsvFile(dataPath).types(StringValue.class, BooleanValue.class, ByteValue.class, ShortValue.class, IntValue.class, LongValue.class, FloatValue.class, DoubleValue.class);
+		data.writeAsText(resultPath);
```
If I'm not mistaken, then we wanted to avoid writing data to disk because this sometimes fails on Travis. Instead we should use `collect` to keep the data in memory.

chiwanpark replied:
@tillrohrmann Yeah, I know that. But to use `collect` instead of writing to disk, we need to change all test methods in `CsvReaderITCase`. Maybe we can cover this in other issue (FLINK-2032).

tillrohrmann replied:
That is true, but it might ease the problem a little bit if newly added tests try to use `collect`. And I doubt that we'll soon find somebody who will take care of this.

chiwanpark replied:
Okay, I'll update `CsvReaderITCase` to use `collect` instead of writing data to disk.

chiwanpark commented:
I fixed the test to use `collect` method.

fhueske commented:
Good to merge

StephanEwen commented:
Merging this...
